url3c955055f0_dirkzhang_antlr_expressionEngine_tgz-pJ8-expression_DirkExpTestJ8/src/expression/DirkExpLexer.java

```diff
@@ -0,0 +1,1195 @@
+            // F:\\antlr\\BoncExp.g:180:5: ( '[' )
+            // F:\\antlr\\BoncExp.g:180:9: '['
+            {
+            match('['); 
+
+            }
+
+            state.type = _type;
+            state.channel = _channel;
+        }
+        finally {
+        }
+    }
+    // $ANTLR end "LBRACKET"
+
+    // $ANTLR start "RBRACKET"
+    public final void mRBRACKET() throws RecognitionException {
+        try {
+            int _type = RBRACKET;
+            int _channel = DEFAULT_TOKEN_CHANNEL;
+            // F:\\antlr\\BoncExp.g:184:5: ( ']' )
+            // F:\\antlr\\BoncExp.g:184:9: ']'
+            {
+            match(']'); 
+
+            }
+
+            state.type = _type;
+            state.channel = _channel;
+        }
+        finally {
+        }
+    }
+    // $ANTLR end "RBRACKET"
+
+    // $ANTLR start "DATETIME_LITERAL"
+    public final void mDATETIME_LITERAL() throws RecognitionException {
+        try {
+            int _type = DATETIME_LITERAL;
+            int _channel = DEFAULT_TOKEN_CHANNEL;
+            // F:\\antlr\\BoncExp.g:188:5: ( '\\'' STRING_LITERAL '\\'' )
+            // F:\\antlr\\BoncExp.g:188:9: '\\'' STRING_LITERAL '\\''
+            {
+            match('\''); 
+            mSTRING_LITERAL(); 
+            match('\''); 
+
+            }
+
+            state.type = _type;
+            state.channel = _channel;
+        }
+        finally {
+        }
+    }
+    // $ANTLR end "DATETIME_LITERAL"
+
+    // $ANTLR start "STRING_LITERAL"
+    public final void mSTRING_LITERAL() throws RecognitionException {
+        try {
+            int _type = STRING_LITERAL;
+            int _channel = DEFAULT_TOKEN_CHANNEL;
+            // F:\\antlr\\BoncExp.g:192:5: ( '\"' (~ ( '\"' | '[' | ']' ) )* '\"' )
+            // F:\\antlr\\BoncExp.g:192:9: '\"' (~ ( '\"' | '[' | ']' ) )* '\"'
+            {
+            match('\"'); 
+            // F:\\antlr\\BoncExp.g:192:13: (~ ( '\"' | '[' | ']' ) )*
+            loop2:
+            do {
+                int alt2=2;
+                int LA2_0 = input.LA(1);
+
+                if ( ((LA2_0>='\u0000' && LA2_0<='!')||(LA2_0>='#' && LA2_0<='Z')||LA2_0=='\\'||(LA2_0>='^' && LA2_0<='\uFFFF')) ) {
+                    alt2=1;
+                }
+
+
+                switch (alt2) {
+            	case 1 :
+            	    // F:\\antlr\\BoncExp.g:192:14: ~ ( '\"' | '[' | ']' )
+            	    {
+            	    if ( (input.LA(1)>='\u0000' && input.LA(1)<='!')||(input.LA(1)>='#' && input.LA(1)<='Z')||input.LA(1)=='\\'||(input.LA(1)>='^' && input.LA(1)<='\uFFFF') ) {
+            	        input.consume();
+
+            	    }
+            	    else {
+            	        MismatchedSetException mse = new MismatchedSetException(null,input);
+            	        recover(mse);
+            	        throw mse;}
+
+
+            	    }
+            	    break;
+
+            	default :
+            	    break loop2;
+                }
+            } while (true);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 567:596, 621:650.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.